share/completions/rustup.fish

```diff
@@ -264,3 +264,34 @@ complete -c rustup -n "__fish_seen_subcommand_from set; and __fish_prev_arg_in p
 # Global argument completions where valid
 complete -c rustup -n "__fish_prev_arg_in --toolchain" -xa "$__rustup_toolchains_short $__rustup_toolchains"
 
+complete -f -c rustup -n '__fish_seen_subcommand_from doc' -a '(__fish_rustup_docs_modules)'
+complete -f -c rustup -n '__fish_seen_subcommand_from doc' -a '(__fish_rustup_docs_keywords)'
+complete -f -c rustup -n '__fish_seen_subcommand_from doc' -a '(__fish_rustup_docs_primitives)'
+complete -f -c rustup -n '__fish_seen_subcommand_from doc' -a '(__fish_rustup_docs_macros)'
+
+function __fish_rustup_docs_modules
+  set -l doc_path (dirname (rustup doc --path))
+  command find "$doc_path"/{std,core,alloc} -name index.html \
+  | string replace --regex "$doc_path/(.+)/index\.html" '$1\tModule' \
+  | string replace --all / ::
+end
+
+function __fish_rustup_docs_keywords
+  set -l doc_path (dirname (rustup doc --path))
+  command ls $doc_path/std/keyword.*.html \
```

Comment:
These `ls` aren't needed, and in fact with fish's globbing behavior will error if nothing matches. Do it like

```fish
set -l doc_paths (dirname (rustup doc --path))/std/keyword.*.html
string replace --regex "$doc_path/std/keyword\.(.+)\.html" '$1\tKeyword' -- $doc_paths
```

instead. Also I strongly suspect the "dirname" is unneeded. What's the output from `rustup doc --path` like?

Reply:
I can't see how it would be needed here - `dirname` removes the last non-slash component and trailing slashes.

So it turns `/usr/share/doc/` into `/usr/share/doc` and `/usr/share/doc/index.html` into `/usr/share/doc`.

But I highly doubt `rustup doc --path` would give a file *inside* the path, so the latter doesn't apply and the former we don't care about because multiple successive slashes are valid in paths - try `ls ///usr///share/doc`. We also don't care about it turning $PWD into `.` or any possible canonicalization, because we don't care about the path to begin with - we only care about files inside it.

Reply:
On my system I get this:
```
/Users/collin.styles/.rustup/toolchains/stable-x86_64-apple-darwin/share/doc/rust/html/index.html
```

Reply:
In your suggestion, you use both `$doc_path` and `$doc_paths`. So would something like this be preferrable?

```shell
function __fish_rustup_docs_keywords
  set -l doc_path (dirname (rustup doc --path))
  set -l doc_paths "$doc_path"/std/keyword.*.html
  string replace --regex "$doc_path/std/keyword\.(.+)\.html" '$1\tKeyword' -- $doc_paths
end
```

Reply:
>In your suggestion, you use both $doc_path and $doc_paths

Yeah, sorry, that was just a typo. It's fine to do it with one variable, but two is also okay.

I wouldn't be calling `dirname` either, because that's an external command, which is often slower and has issues with portability (tho just calling `dirname` is probably supported "everywhere").

 I'd be using our `string replace`, like `rustup doc --path | string replace --regex '/[^/]*$' ''` - remove everything after the last slash.